### PR TITLE
FIX: Callback ordering issues when adding/removing during execution (case 1322530).

### PR DESCRIPTION
**Pull request by Rene-Damm**
Fixes [1322530](https://issuetracker.unity3d.com/issues/inputsystems-events-are-not-called-the-order-they-were-added-when-they-are-modified-in-the-middle-of-the-call-by-other-listener) ([FogBugz](https://fogbugz.unity3d.com/f/cases/1322530/)).

### Description

Adding and removing callbacks from within callbacks isn't handled properly in the code. Some stuff in `DelegateHelpers` but falls over sideways quickly. Has come up a number of times in reports. There was a user PR about this, too (wasn't the right fix).

### Changes made

Bit the bullet and added a `CallbackArray` version that made a copy of the list before going through callbacks.

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [x] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs [hunk 1]
```diff
@@ -1055,8 +1002,7 @@ private void NotifyUsageChanged(InputDevice device)
             InputActionState.OnDeviceChange(device, InputDeviceChange.UsageChanged);
 
             // Notify listeners.
-            for (var i = 0; i < m_DeviceChangeListeners.length; ++i)
-                m_DeviceChangeListeners[i](device, InputDeviceChange.UsageChanged);
+            DelegateHelpers.InvokeCallbacksSafe(ref m_DeviceChangeListeners, device, InputDeviceChange.UsageChanged, "InputSystem.onDeviceChange");
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/InputManager.cs:1006]: I know the last argument here is just for logging and profiling, but is it right? They're all onDeviceChange?

**Review by andrew-oc — COMMENTED** [review]
A neat solution. I'm wondering about the PrepareExecution method and paying the cost of the array copy (potentially, depending on number of events) whether the delegates are modified or not. Did you consider as an alternative changing CallbackArray so it sets a flag when delegates are being enumerated and any delegates that are added or removed while the flag is active get added to a different buffer maintained by CallbackArray. When enumeration finishes, any additions or removals get added into the main array. Basically you only pay the small performance hit when you actually change the delegates.

**Comment by Rene-Damm** [comment]
> A neat solution. I'm wondering about the PrepareExecution method and paying the cost of the array copy (potentially, depending on number of events) whether the delegates are modified or not. Did you consider as an alternative changing CallbackArray so it sets a flag when delegates are being enumerated and any delegates that are added or removed while the flag is active get added to a different buffer maintained by CallbackArray. When enumeration finishes, any additions or removals get added into the main array. Basically you only pay the small performance hit when you actually change the delegates.

Did consider it but for some reason deemed it the worse option. Actually tried it out now and doesn't look that bad. Let me know what you think.

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs (hunk 1 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/InputSystem/InputManager.cs:1006]: Yup, these are correct. It's all just different InputDeviceChange enum values going out on the same `onDeviceChange` API endpoint.

**Review by andrew-oc — APPROVED** [review]

**Pull request merged**